File: Algorithms/C++/Easy/257_Binary_Tree_Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> paths;

        helper(paths, "", root);

        return paths;
    }
private:
    void helper(vector<string>& paths, string path, TreeNode* root) {
        if (root == NULL) {
            return;
        }

        path += to_string(root->val);

        if (root->left == NULL && root->right == NULL) {
            paths.push_back(path);
            return;
        }

        path += "->";
        helper(paths, path, root->left);
        helper(paths, path, root->right);
    }
};

File: Algorithms/JavaScript/Easy/257_Binary_Tree_Paths.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {string[]}
 */
var binaryTreePaths = function(root) {
    var paths = []

    var currentPathValues = []
    helper(root, currentPathValues, paths)

    return paths
};

var helper = function(root, currentPathValues, paths) {
    if (root === null) {
        return
    }

    currentPathValues.push(root.val)

    if (root.left === null && root.right === null) {
        paths.push(generatePath(currentPathValues))
    } else {
        helper(root.left, currentPathValues, paths)
        helper(root.right, currentPathValues, paths)
    }

    currentPathValues.pop()
}

var generatePath = function(pathValues) {
    if (pathValues.length === 0) {
        return ""
    }

    var path = ""
    for (var i = 0; i < pathValues.length - 1; i++) {
        path += pathValues[i] + "->"
    }

    path += pathValues[pathValues.length - 1]

    return path
}
